internal/jimm/role/role_test.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2024 Canonical.
+package role
+
+import (
+	"context"
+	"sort"
+	"strconv"
+	"testing"
+	"time"
+
+	cofga "github.com/canonical/ofga"
+	qt "github.com/frankban/quicktest"
+	"github.com/frankban/quicktest/qtsuite"
+
+	"github.com/canonical/jimm/v3/internal/common/pagination"
+	"github.com/canonical/jimm/v3/internal/db"
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/openfga"
+	ofganames "github.com/canonical/jimm/v3/internal/openfga/names"
+	jimmtest "github.com/canonical/jimm/v3/internal/testutils/jimmtest"
+)
+
+type roleManagerSuite struct {
+	manager    *roleManager
+	user       *openfga.User
+	db         *db.Database
+	ofgaClient *openfga.OFGAClient
+}
+
+func (s *roleManagerSuite) Init(c *qt.C) {
+	// Setup DB
+	db := &db.Database{
+		DB: jimmtest.PostgresDB(c, time.Now),
+	}
+	err := db.Migrate(context.Background(), false)
+	c.Assert(err, qt.IsNil)
+
+	s.db = db
+
+	// Setup OFGA
+	ofgaClient, _, _, err := jimmtest.SetupTestOFGAClient(c.Name())
+	c.Assert(err, qt.IsNil)
+
+	s.ofgaClient = ofgaClient
+
+	s.manager, err = NewRoleManager(db, ofgaClient)
+	c.Assert(err, qt.IsNil)
+
+	// Create test identity
+	i, err := dbmodel.NewIdentity("alice")
+	c.Assert(err, qt.IsNil)
+	s.user = openfga.NewUser(i, ofgaClient)
+}
+
+func (s *roleManagerSuite) TestAddRole(c *qt.C) {
+	c.Parallel()
+	ctx := context.Background()
+
+	// Check only admin can AddRole
+	_, err := s.manager.AddRole(ctx, s.user, "models")
+	c.Assert(errors.ErrorCode(err), qt.Equals, errors.CodeUnauthorized)
+
+	// Check happypath
+	s.user.JimmAdmin = true
+
+	re, err := s.manager.AddRole(ctx, s.user, "models")
+	c.Assert(err, qt.IsNil)
+
+	c.Assert(re, jimmtest.DBObjectEquals, &dbmodel.RoleEntry{
+		Name: "models",
+		UUID: re.UUID,
+		ID:   1,
+	})
+}
+
+func (s *roleManagerSuite) TestGetRole(c *qt.C) {
+	c.Parallel()
+	ctx := context.Background()
+
+	// Check only admin can GetRoleByName (this affects UUID too)
+	_, err := s.manager.GetRoleByName(ctx, s.user, "models")
+	c.Assert(errors.ErrorCode(err), qt.Equals, errors.CodeUnauthorized)
+
+	// Check happypath
+	s.user.JimmAdmin = true
+
+	re, err := s.db.AddRole(ctx, "models")
+	c.Assert(err, qt.IsNil)
+
+	re2, err := s.manager.GetRoleByUUID(ctx, s.user, re.UUID)
+	c.Assert(err, qt.IsNil)
+	c.Assert(re.ID, qt.Equals, re2.ID)
+
+	re3, err := s.manager.GetRoleByName(ctx, s.user, re.Name)
+	c.Assert(err, qt.IsNil)
+	c.Assert(re.ID, qt.Equals, re3.ID)
+}
+
+func (s *roleManagerSuite) TestRemoveRole(c *qt.C) {
+	c.Parallel()
+	ctx := context.Background()
+
+	// Check only admin can RemoveRole
+	err := s.manager.RemoveRole(ctx, s.user, "r1")
+	c.Assert(errors.ErrorCode(err), qt.Equals, errors.CodeUnauthorized)
+
+	// Check happypath
+	s.user.JimmAdmin = true
+
+	// Add to the db and openfga.
+	re, err := s.db.AddRole(ctx, "r1")
+	c.Assert(err, qt.IsNil)
+
+	tupe := openfga.Tuple{
```

Comment:
-> `tuple`

Reply:
I really love the word "tupe" I can't deny aha